internal/wasmldr/src/loader/compiled/tls.rs

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::any::Any;
+use std::io::{ErrorKind, IoSlice, IoSliceMut, Read, Write};
+use std::sync::Arc;
+
+use rustix::fd::AsFd;
+use rustls::{ClientConfig, ClientConnection, Connection, ServerConfig, ServerConnection};
+use wasi_common::file::{FdFlags, FileType};
+use wasi_common::{Context, Error, ErrorExt, WasiFile};
+
+pub struct Stream {
+    tcp: std::net::TcpStream,
+    tls: Connection,
+}
+
+impl From<Stream> for Box<dyn WasiFile> {
+    fn from(value: Stream) -> Self {
+        Box::new(value)
+    }
+}
+
+impl Stream {
+    pub fn connect(
+        mut tcp: std::net::TcpStream,
+        name: &str,
+        cfg: Arc<ClientConfig>,
+    ) -> Result<Self, Error> {
+        // Set up connection.
+        let tls = ClientConnection::new(cfg, name.try_into()?)?;
+        let mut tls = Connection::Client(tls);
+
+        // Finish the connection.
+        tls.complete_io(&mut tcp)?;
+
+        Ok(Self { tcp, tls })
+    }
+}
+
+#[wiggle::async_trait]
+impl WasiFile for Stream {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    async fn get_filetype(&mut self) -> Result<FileType, Error> {
+        Ok(FileType::SocketStream)
+    }
+
+    async fn get_fdflags(&mut self) -> Result<FdFlags, Error> {
+        Ok(FdFlags::APPEND)
+    }
+
+    async fn read_vectored<'a>(&mut self, bufs: &mut [IoSliceMut<'a>]) -> Result<u64, Error> {
+        if self.tls.wants_read() {
+            self.tls.read_tls(&mut self.tcp)?;
+            self.tls.process_new_packets()?;
+        }
+
+        let n = match self.tls.reader().read_vectored(bufs) {
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => 0,
+            Err(e) => return Err(e.into()),
+            Ok(n) => n,
+        };
+
+        Ok(n as u64)
```

Comment:
Why cast and not `.try_into()` conversion as in original draft?

Reply:
We can had `.try_into()` if you want. But all conversions from `usize` to `u64` are infallible on all existing hardware platforms. And I don't see us getting 128bit CPUs anytime soon.